typesetters/base.lua

```diff
@@ -315,18 +322,147 @@ function typesetter:breakIntoLines (nodelist, breakWidth)
   return self:breakpointsToLines(breakpoints)
 end
 
-function typesetter.shapeAllNodes (_, nodelist)
-  local newNl = {}
-  for i = 1, #nodelist do
-    if nodelist[i].is_unshaped then
-      pl.tablex.insertvalues(newNl, nodelist[i]:shape())
+local function getLastShape(nodelist)
+  local last
+  if nodelist then
+    -- The node list may contain nnodes, penalties, kern and glue
+    -- We skip the latter, and retrieve the last shaped item.
+    for i = #nodelist, 1, -1 do
+      local n = nodelist[i]
+      if n.is_nnode then
+        local items = n.nodes[#n.nodes].value.items
+        last = items[#items]
+        break
+      end
+      if n.is_kern and n.subtype == "punctspace" then
+        -- Some languages such as French insert a special space around
+        -- punctuations. In those case, we should not need italic correction.
```

Comment:
> In those case, we should not need italic correction.

@sile-typesetter/lang-fr Note that is is debatable, and I couldn't find anything about it in typography resources I can access to... I hesitated between:
- No italic correction (the current choice here)
- Italic correction always applied, but it looked really exaggerate with most fonts I tried... and I mean even reasonable choices (e.g. EB Garamond).
- Adding an additional setting so as to define a ratio to apply to the italic correction (0..1) for maximum flexibility... But it sounded far-fetched for that small niche language... and not many words in French end with an _f_... Er. Ok, _turf_ does, for instance. It's not even really French. And it doesn't look that bad anyway...)

Reply:
To better explain:
The current implementation cancels italic correction in the cases I marked in red below.
![image](https://user-images.githubusercontent.com/18075640/235312438-d7ed1170-3aef-4543-9b2a-0ffad1b77e41.png)

Seems to be the french punctuation spacing does not need them indeed.